src/validators/model_class.rs

```diff
@@ -100,50 +100,43 @@ impl Validator for ModelClassValidator {
 
 impl ModelClassValidator {
     unsafe fn create_class(&self, py: Python, output: PyObject) -> PyResult<PyObject> {
-        let t: &PyTuple = output.extract(py)?;
-        let model_dict = t.get_item(0)?;
-        let fields_set = t.get_item(1)?;
+        let (model_dict, fields_set): (&PyAny, &PyAny) = output.extract(py)?;
 
         // based on the following but with the second argument of new_func set to an empty tuple as required
         // https://github.com/PyO3/pyo3/blob/d2caa056e9aacc46374139ef491d112cb8af1a25/src/pyclass_init.rs#L35-L77
         let args = PyTuple::empty(py);
         let raw_type = self.class.as_ref(py).as_type_ptr();
-        let instance_ptr = match (*raw_type).tp_new {
-            Some(new_func) => {
-                let obj = new_func(raw_type, args.as_ptr(), null_mut());
-                if obj.is_null() {
-                    return Err(PyErr::fetch(py));
-                } else {
-                    obj
-                }
-            }
+        let instance = match (*raw_type).tp_new {
+            Some(new_func) => PyObject::from_owned_ptr_or_err(py, new_func(raw_type, args.as_ptr(), null_mut()))?,
             None => return Err(PyTypeError::new_err("base type without tp_new")),
         };
 
-        force_setattr(instance_ptr, py, intern!(py, "__dict__"), model_dict)?;
-        force_setattr(instance_ptr, py, intern!(py, "__fields_set__"), fields_set)?;
+        let instance_ref = instance.as_ref(py);
+        force_setattr(instance_ref, intern!(py, "__dict__"), model_dict)?;
+        force_setattr(instance_ref, intern!(py, "__fields_set__"), fields_set)?;
 
-        match PyAny::from_borrowed_ptr_or_opt(py, instance_ptr) {
-            Some(instance) => Ok(instance.into()),
-            None => Err(PyTypeError::new_err("failed to create instance of class")),
-        }
+        Ok(instance)
     }
 }
 
 /// copied and modified from
 /// https://github.com/PyO3/pyo3/blob/d2caa056e9aacc46374139ef491d112cb8af1a25/src/instance.rs#L587-L597
 /// to use `PyObject_GenericSetAttr` thereby bypassing `__setattr__` methods on the instance,
 /// see https://github.com/PyO3/pyo3/discussions/2321 for discussion
-pub fn force_setattr<N, V>(obj: *mut ffi::PyObject, py: Python<'_>, attr_name: N, value: V) -> PyResult<()>
+pub fn force_setattr<N, V>(obj: &PyAny, attr_name: N, value: V) -> PyResult<()>
 where
     N: ToPyObject,
     V: ToPyObject,
 {
-    attr_name.with_borrowed_ptr(py, move |attr_name| {
-        value.with_borrowed_ptr(py, |value| unsafe {
-            error_on_minusone(py, ffi::PyObject_GenericSetAttr(obj, attr_name, value))
-        })
-    })
+    let py = obj.py();
```

Comment:
is there any reason to define `py` here rather than passing it as an argument?

I tend to pass it around as an argument because I had assumed it was marginally more performant, but I never actually checked.

Reply:
I think in both cases it should get completely optimised out; `Python` is a zero-sized type and `obj.py()` is an inline function with effectively an empty body. All `Python<'_>` is used for is to carry around that promise of the GIL lifetime in the type system.

Will change to match the style you've used elsewhere.